metrics/sinks/poseidon/driver.go

```diff
@@ -0,0 +1,143 @@
+package poseidon
+
+import (
+	"fmt"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"k8s.io/heapster/metrics/core"
+	"k8s.io/heapster/metrics/sinks/poseidon/stats"
+	"log"
+	"net/url"
+	"sync"
+)
+
+type PoseidonSink struct {
+	client stats.PoseidonStatsClient
+	sync.RWMutex
+}
+
+func (sink *PoseidonSink) Name() string {
+	return "Poseidon Sink"
+}
+
+func (sink *PoseidonSink) Stop() {
+	// Do nothing.
+}
+
+func (sink *PoseidonSink) ExportData(dataBatch *core.DataBatch) {
+	sink.Lock()
+	defer sink.Unlock()
+
+	for metricKey, metricSet := range dataBatch.MetricSets {
+		if isNodeMetric(metricKey, metricSet) {
+			nodeStats := &stats.NodeStats{
+				Hostname:       metricSet.Labels[core.LabelHostname.Key],
+				CpuAllocatable: uint64(getMetricValue(metricSet, core.MetricNodeCpuAllocatable).FloatValue),
+				CpuCapacity:    uint64(getMetricValue(metricSet, core.MetricNodeCpuCapacity).FloatValue),
+				CpuReservation: uint64(getMetricValue(metricSet, core.MetricNodeCpuReservation).FloatValue),
+				CpuUtilization: uint64(getMetricValue(metricSet, core.MetricNodeCpuUtilization).FloatValue),
+				MemAllocatable: uint64(getMetricValue(metricSet, core.MetricNodeMemoryAllocatable).FloatValue),
+				MemCapacity:    uint64(getMetricValue(metricSet, core.MetricNodeMemoryCapacity).FloatValue),
+				MemReservation: uint64(getMetricValue(metricSet, core.MetricNodeMemoryReservation).FloatValue),
+				MemUtilization: uint64(getMetricValue(metricSet, core.MetricNodeMemoryUtilization).FloatValue),
+			}
+			stream, err := sink.client.ReceiveNodeStats(context.Background())
+			if err != nil {
+				log.Fatalf("Error connecting the server: %v", err)
+			}
+			nodeStatsResp, err := stream.Recv()
+			if err != nil {
+				log.Fatalf("Error receiving node stats response: %v", err)
+			}
+			switch nodeStatsResp.Type {
+			case stats.NodeStatsResponseType_NODE_NOT_FOUND:
+				log.Fatalf("Node %s not found", nodeStatsResp.GetHostname())
+			case stats.NodeStatsResponseType_NODE_STATS_OK:
+				err := stream.Send(nodeStats)
+				if err != nil {
+					log.Fatalf("Error sending node stats: %v", err)
+				}
+			default:
+				panic(fmt.Sprintf("Unexpected node stats response: %v", err))
+			}
+		} else if isPodMetric(metricKey, metricSet) {
+			podStats := &stats.PodStats{
+				Name:      metricSet.Labels[core.LabelPodName.Key],
+				Namespace: metricSet.Labels[core.LabelPodNamespace.Key],
+				Hostname:  metricSet.Labels[core.LabelHostname.Key],
+				// CPU stats
+				CpuLimit:   uint64(getMetricValue(metricSet, core.MetricCpuLimit).IntValue),
+				CpuRequest: uint64(getMetricValue(metricSet, core.MetricCpuRequest).IntValue),
+				CpuUsage:   uint64(getMetricValue(metricSet, core.MetricCpuUsage).IntValue),
+				// Memory stats
+				MemLimit:            uint64(getMetricValue(metricSet, core.MetricMemoryLimit).IntValue),
+				MemRequest:          uint64(getMetricValue(metricSet, core.MetricMemoryRequest).IntValue),
```

Comment:
Some of the memory metric units are counts, like page faults and rates. Those should stay as is, right? Also, Some of the network metrics are in bytes, too, like NetRx and NetTx. These should be converted to KB, too. Correct?